clients/java/src/main/java/io/zeebe/client/impl/OAuthCredentialsProvider.java

```diff
@@ -0,0 +1,122 @@
+import java.io.OutputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class OAuthCredentialsProvider implements CredentialsProvider {
+  private static final ObjectMapper MAPPER = new ObjectMapper();
+  private static final ObjectReader JSON_READER = MAPPER.readerFor(ZeebeClientCredentials.class);
+  private static final Logger LOG = LoggerFactory.getLogger(OAuthCredentialsProvider.class);
+  private static final Key<String> HEADER_AUTH_KEY =
+      Key.of("Authorization", Metadata.ASCII_STRING_MARSHALLER);
+
+  private final URL authorizationServerUrl;
+  private final String jsonPayload;
+
+  private ZeebeClientCredentials credentials;
+
+  public OAuthCredentialsProvider(OAuthCredentialsProviderBuilder builder) {
+    authorizationServerUrl = builder.getAuthorizationServer();
+    jsonPayload = createJsonPayload(builder);
+  }
+
+  /** Adds an access token to the Authorization header of a gRPC call. */
+  @Override
+  public void applyCredentials(Metadata headers) {
+    try {
+      final ZeebeClientCredentials credentials = getCredentials();
+
+      headers.put(
+          HEADER_AUTH_KEY,
+          String.format(
+              "%s %s", credentials.getTokenType().trim(), credentials.getAccessToken().trim()));
+    } catch (IOException e) {
+      LOG.warn("Failed while fetching credentials, will not add credentials to rpc: ", e);
+    }
+  }
+
+  private ZeebeClientCredentials getCredentials() throws IOException {
+    if (credentials == null) {
+      credentials = fetchCredentials();
+    }
+
+    return credentials;
+  }
+
+  private static String createJsonPayload(OAuthCredentialsProviderBuilder builder) {
+    try {
+      final Map<String, String> payload =
+          new HashMap<String, String>() {
+            {
+              put("client_id", builder.getClientId());
+              put("client_secret", builder.getClientSecret());
+              put("audience", builder.getAudience());
+              put("grant_type", "client_credentials");
+            }
+          };
+
+      return MAPPER.writeValueAsString(payload);
+    } catch (JsonProcessingException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private ZeebeClientCredentials fetchCredentials() throws IOException {
+    final HttpURLConnection connection =
+        (HttpURLConnection) authorizationServerUrl.openConnection();
+    connection.setRequestMethod("POST");
+    connection.setRequestProperty("Content-Type", "application/json");
+    connection.setRequestProperty("Accept", "application/json");
+    connection.setDoOutput(true);
+
+    try (OutputStream os = connection.getOutputStream()) {
+      final byte[] input = jsonPayload.getBytes(StandardCharsets.UTF_8);
+      os.write(input, 0, input.length);
+    }
+
+    if (connection.getResponseCode() != 200) {
```

Comment:
Are we sure it's always 200? What if we got a 201?

Reply:
What the OAuth 2.0 RFC says about the  "constructs the response by adding the following parameters to the entity-body of the HTTP response with a 200 (OK) status code: (...)" https://tools.ietf.org/html/rfc6749#section-5.1